jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java

```diff
@@ -0,0 +1,735 @@
+import org.jgrapht.graph.*;
+import org.jgrapht.nio.*;
+import org.jgrapht.util.*;
+
+/**
+ * Importer for files in the
+ * <a href="http://comopt.ifi.uni-heidelberg.de/software/TSPLIB95/">TSPLIB95</a> format.
+ * <p>
+ * This importer reads the nodes of a <em>Symmetric travelling salesman problem</em> instance from a
+ * file and creates a {@link GraphTests#isComplete(Graph) complete graph} and provides further data
+ * from the file and about the imported graph.
+ * </p>
+ * <p>
+ * This implementation does not cover the full TSPLIB95 standard and only implements the subset of
+ * capabilities required at the time of creation. The following keywords of <em>The specification
+ * part</em> in chapter 1.1 of the <em>TSPLIB95</em> standard are supported:
+ * <ul>
+ * <li>NAME</li>
+ * <li>TYPE</li>
+ * <li>COMMENT</li>
+ * <li>DIMENSION</li>
+ * <li>EDGE_WEIGHT_TYPE</li> Only the following edge weight types are supported
+ * <ul>
+ * <li>EUC_2D</li>
+ * <li>EUC_3D</li>
+ * <li>MAX_2D</li>
+ * <li>MAX_3D</li>
+ * <li>MAN_2D</li>
+ * <li>MAN_3D</li>
+ * <li>CEIL2D</li>
+ * <li>GEO</li>
+ * <li>ATT</li>
+ * </ul>
+ * </ul>
+ * The values of all supported keywords can be obtained from corresponding getters.
+ * </p>
+ * <p>
+ * The following data sections of <em>The data part</em> in chapter 1.2 of the <em>TSPLIB95</em>
+ * standard are supported:
+ * <ul>
+ * <li>NODE_COORD_SECTION</li>
+ * </ul>
+ * </p>
+ * <p>
+ * It was attempted to make the structure of this implementation generic so further keywords from
+ * the specification part or other data sections can be considered if required by broaden this
+ * class. Currently this implementation only reads <em>Symmetric travelling salesman problems</em>
+ * with a NODE_COORD_SECTION and on of the supported EDGE_WEIGHT_TYPE.
+ * </p>
+ * <p>
+ * The website of the TSPLIB standard already contains a large library of different TSP instances
+ * provided as files in TSPLIB format. The
+ * <a href="http://www.math.uwaterloo.ca/tsp/data/index.html">TSPLIB library of the University of
+ * Waterlo</a> provides more problem instances, among others a World TSP and instances based on
+ * cities of different countries.
+ * </p>
+ * 
+ * @author Hannes Wellmann
+ * @param <V> the type of vertices created by this importer
+ *
+ */
+public class TSPLIBImporter<V>
+    implements
+    GraphImporter<V, DefaultWeightedEdge>
+{
+    /**
+     * Container for data of an imported <em>TSPLIB95</em> file.
+     * 
+     * @author Hannes Wellmann
+     * @param <V> the type of vertices in the imported graph
+     */
+    public static class TSPLIBFileData<V>
+    {
+
+        private String name;
+        private String type;
+        private String comment;
+        private Integer dimension;
+        private String edgeWeightType;
+        private String nodeCoordType;
+        private boolean hasDistinctLocations;
+        private Boolean hasDistinctEdges;
+
+        private Graph<V, DefaultWeightedEdge> graph;
```

Comment:
If you are only using this for `hasDistinctEdgesPerVertex` then ok, otherwise this is not needed.
The user provides the graph from the `importGraph` method, no need to include it back.